ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/trielog/TrieLogPruner.java

```diff
@@ -0,0 +1,186 @@
+import com.google.common.collect.TreeMultimap;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TrieLogPruner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TrieLogPruner.class);
+
+  public static final long DEFAULT_RETENTION_THRESHOLD = 0L;
+  public static int DEFAULT_PRUNING_LIMIT = 30_000;
+
+  private final int pruningLimit;
+  private final int loadingLimit;
+  private final BonsaiWorldStateKeyValueStorage rootWorldStateStorage;
+  private final Blockchain blockchain;
+  private final long numBlocksToRetain;
+  private static final Multimap<Long, byte[]> knownTrieLogKeysByDescendingBlockNumber =
+      TreeMultimap.create(Comparator.reverseOrder(), Comparator.comparingInt(Arrays::hashCode));
+
+  public TrieLogPruner(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final Blockchain blockchain,
+      final long numBlocksToRetain,
+      final int pruningLimit) {
+    this.rootWorldStateStorage = rootWorldStateStorage;
+    this.blockchain = blockchain;
+    this.numBlocksToRetain = numBlocksToRetain;
+    this.pruningLimit = pruningLimit;
+    this.loadingLimit = pruningLimit; // same as pruningLimit for now
+  }
+
+  public void initialize() {
+    preloadCache();
+  }
+
+  private void preloadCache() {
+    LOG.atInfo()
+        .setMessage("Loading first {} trie logs from database...")
+        .addArgument(loadingLimit)
+        .log();
+    final Stream<byte[]> trieLogKeys = rootWorldStateStorage.streamTrieLogKeys(loadingLimit);
+    final AtomicLong count = new AtomicLong();
+    trieLogKeys.forEach(
+        blockHashAsBytes -> {
+          Hash hash = Hash.wrap(Bytes32.wrap(blockHashAsBytes));
+          final Optional<BlockHeader> header = blockchain.getBlockHeader(hash);
+          if (header.isPresent()) {
+            knownTrieLogKeysByDescendingBlockNumber.put(header.get().getNumber(), blockHashAsBytes);
+            count.getAndIncrement();
+          } else {
+            // prune orphaned blocks (sometimes created during block production)
+            rootWorldStateStorage.pruneTrieLog(blockHashAsBytes);
+          }
+        });
+    LOG.atInfo().log("Loaded {} trie logs from database", count);
+  }
+
+  void cacheForLaterPruning(final long blockNumber, final byte[] trieLogKey) {
+    LOG.atTrace()
+        .setMessage("caching trie log for later pruning blockNumber {}; trieLogKey (blockHash) {}")
+        .addArgument(blockNumber)
+        .addArgument(Bytes.wrap(trieLogKey).toHexString())
+        .log();
+    knownTrieLogKeysByDescendingBlockNumber.put(blockNumber, trieLogKey);
+  }
+
+  void pruneFromCache() {
+
+    final long retainAboveThisBlock = blockchain.getChainHeadBlockNumber() - numBlocksToRetain;
+    final long retainAboveThisBlockOrFinalized =
+        blockchain
+            .getFinalized()
+            .flatMap(blockchain::getBlockHeader)
```

Comment:
Maybe could save a db call if we combine into one transaction

Reply:
why not using the same logic as snapsync, https://github.com/hyperledger/besu/blob/6a2d41fccc00eac110a8cec27bc16cc2df47da0a/besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java#L860
I think we are not calling the database with this method, maybe only the first time

Reply:
thanks, that might be a nice way to avoid the db...will look into it